src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
             {
-                Code = diagnostic.Id,
-                Message = diagnostic.Message,
-                Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnostic.Severity),
-                Range = ProtocolConversions.TextSpanToRange(DiagnosticData.GetExistingOrCalculatedTextSpan(diagnostic.DataLocation, text), text),
-                // Only the unnecessary diagnostic tag is currently supported via LSP.
-                Tags = diagnostic.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
-            }).ToArray();
+                Contract.ThrowIfNull(diagnosticData.DataLocation, "Diagnostic data location should not be null here");
+
+                var filePath = diagnosticData.DataLocation.MappedFilePath ?? diagnosticData.DataLocation.OriginalFilePath;
+                return ProtocolConversions.GetUriFromFilePath(filePath);
+            }
+
+            static LanguageServer.Protocol.Diagnostic ConvertToLspDiagnostic(DiagnosticData diagnosticData, SourceText text)
+            {
+                return new LanguageServer.Protocol.Diagnostic
+                {
+                    Code = diagnosticData.Id,
+                    Message = diagnosticData.Message,
+                    Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnosticData.Severity),
+                    Range = GetDiagnosticRange(diagnosticData.DataLocation, text),
+                    // Only the unnecessary diagnostic tag is currently supported via LSP.
+                    Tags = diagnosticData.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary)
+                        ? new DiagnosticTag[] { DiagnosticTag.Unnecessary }
+                        : Array.Empty<DiagnosticTag>()
+                };
+            }
+        }
+
+        private bool IncludeDiagnostic(DiagnosticData diagnostic)
+        {
+            if (!diagnostic.Properties.TryGetValue(nameof(DocumentPropertiesService.DiagnosticsLspClientName), out var diagnosticClientName))
+            {
+                // This diagnostic is not restricted to a specific LSP client, so just pass it through
+                return true;
+            }
+
+            // We only include this diagnostic if it directly matches our name.
+            return diagnosticClientName == _clientName;
+        }
+
+        private static LanguageServer.Protocol.Range? GetDiagnosticRange(DiagnosticDataLocation? diagnosticDataLocation, SourceText text)
+        {
+            var linePositionSpan = DiagnosticData.GetLinePositionSpan(diagnosticDataLocation, text, useMapped: true);
+            return ProtocolConversions.LinePositionToRange(linePositionSpan);
+        }
+
+        private static V GetOrValue<K, V>(Dictionary<K, V> dict, K key, V defaultValue)
+        {
+            if (dict.TryGetValue(key, out var immutableArray))
```

Comment:
```suggestion
            if (dict.TryGetValue(key, out var value))
```

Field name doesn't make sense otherwise. Do you also want to add it this to https://github.com/dotnet/roslyn/blob/master/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/IDictionaryExtensions.cs instead?